File: ui-automation-framework/pages/product-details.page.ts
import type { Locator, Page } from '@playwright/test';
import { HeaderComponent } from '../components/header.component';

export class ProductDetailsPage {
  readonly header: HeaderComponent;
  readonly name: Locator;
  readonly description: Locator;
  readonly price: Locator;
  readonly addToCartButton: Locator;
  readonly backToProductsButton: Locator;

  constructor(readonly page: Page) {
    this.header = new HeaderComponent(page);
    this.name = page.getByTestId('inventory-item-name');
    this.description = page.getByTestId('inventory-item-desc');
    this.price = page.getByTestId('inventory-item-price');
    this.addToCartButton = page.getByTestId('add-to-cart');
    this.backToProductsButton = page.getByTestId('back-to-products');
  }

  async backToProducts(): Promise<void> {
    await this.backToProductsButton.click();
  }
}

File: ui-tests/src/test/java/saucedemo/pages/ProductDetailsPage.java
package saucedemo.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import saucedemo.components.Header;

public class ProductDetailsPage {
    private final Page page;
    public final Header header;

    public ProductDetailsPage(Page page) {
        this.page = page;
        this.header = new Header(page);
    }

    public Locator name() {
        return page.locator("[data-test='inventory-item-name']");
    }

    public Locator description() {
        return page.locator("[data-test='inventory-item-desc']");
    }

    public Locator price() {
        return page.locator("[data-test='inventory-item-price']");
    }

    public Locator addButton() {
        return page.locator("[data-test='add-to-cart']");
    }

    public void addToCart() {
        addButton().click();
    }

    public void backToProducts() {
        page.locator("[data-test='back-to-products']").click();
    }
}
